### ISSUE TITLE
CT: sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/projectile-data-lab#329]
In both PDL and PSD:

```
projectile-data-lab : phet-io-state-fuzz : unbuilt
http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/phet-io-wrappers/state/?sim=projectile-data-lab&locales=*&phetioWrapperDebug=true&fuzz&phetioDebug=true&wrapperContinuousTest=%7B%22test%22%3A%5B%22projectile-data-lab%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1715368625981%22%2C%22timestamp%22%3A1715369403666%7D
Uncaught Error: Assertion failed: sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase. Projectiles in selected sample: 0. Sample size: 2
Error: Assertion failed: sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase. Projectiles in selected sample: 0. Sample size: 2
at window.assertions.assertFunction (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/assert/js/assert.js:28:13)
at assert (SamplingField.ts:377:18)
at step (SamplingModel.ts:282:29)
at step (Sim.ts:432:21)
at apply (PhetioAction.ts:162:16)
at execute (Sim.ts:1048:30)
at stepSimulation (Sim.ts:1038:11)
at stepOneFrame (Sim.ts:1013:11)
[URL] http://128.138.93.172/continuous-testing/aqua/html/wrapper-test.html?url=..%2F..%2Fct-snapshots%2F1715368625981%2Fphet-io-wrappers%2Fstate%2F%3Fsim%3Dprojectile-data-lab%26locales%3D*%26phetioWrapperDebug%3Dtrue%26fuzz%26phetioDebug%3Dtrue&testInfo=%7B%22test%22%3A%5B%22projectile-data-lab%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1715368625981%22%2C%22timestamp%22%3A1715369403666%7D
[NAVIGATED] http://128.138.93.172/continuous-testing/aqua/html/wrapper-test.html?url=..%2F..%2Fct-snapshots%2F1715368625981%2Fphet-io-wrappers%2Fstate%2F%3Fsim%3Dprojectile-data-lab%26locales%3D*%26phetioWrapperDebug%3Dtrue%26fuzz%26phetioDebug%3Dtrue&testInfo=%7B%22test%22%3A%5B%22projectile-data-lab%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1715368625981%22%2C%22timestamp%22%3A1715369403666%7D
[ATTACHED]
[NAVIGATED] about:blank
[NAVIGATED] http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/phet-io-wrappers/state/?sim=projectile-data-lab&locales=*&phetioWrapperDebug=true&fuzz&phetioDebug=true&wrapperContinuousTest=%7B%22test%22%3A%5B%22projectile-data-lab%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1715368625981%22%2C%22timestamp%22%3A1715369403666%7D
[ATTACHED]
[NAVIGATED] about:blank
[ATTACHED]
[NAVIGATED] about:blank
[CONSOLE] enabling assert
[NAVIGATED] http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/projectile-data-lab/projectile-data-lab_en.html?brand=phet-io&ea&postMessageOnError&sim=projectile-data-lab&locales=*&phetioWrapperDebug=true&fuzz&phetioDebug=true&wrapperContinuousTest=%7B%22test%22%3A%5B%22projectile-data-lab%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1715368625981%22%2C%22timestamp%22%3A1715369403666%7D&frameTitle=source
[NAVIGATED] http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/projectile-data-lab/projectile-data-lab_en.html?brand=phet-io&ea&postMessageOnError&sim=projectile-data-lab&locales=*&phetioWrapperDebug=true&fuzz&phetioDebug=true&wrapperContinuousTest=%7B%22test%22%3A%5B%22projectile-data-lab%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1715368625981%22%2C%22timestamp%22%3A1715369403666%7D&frameTitle=destination
[CONSOLE] enabling assert
[CONSOLE] enabling assert
[CONSOLE] continuous-test-wrapper-load
[CONSOLE] continuous-test-wrapper-load
[CONSOLE] Assertion failed: sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase. Projectiles in selected sample: 0. Sample size: 2
[PAGE ERROR] Error: Error: Assertion failed: sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase. Projectiles in selected sample: 0. Sample size: 2
at window.assertions.assertFunction (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/assert/js/assert.js:28:13)
at SamplingField.step (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/projectile-data-lab/js/sampling/model/SamplingField.js:301:19)
at SamplingModel.step (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/projectile-data-lab/js/sampling/model/SamplingModel.js:219:30)
at http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:353:22
at PhetioAction.execute (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/tandem/js/PhetioAction.js:137:17)
at Sim.stepSimulation (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:884:31)
at Sim.stepOneFrame (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:874:12)
at Sim.runAnimationLoop (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:852:12)
[PAGE ERROR] Error: Error: Assertion failed: sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase. Projectiles in selected sample: 0. Sample size: 2
at window.assertions.assertFunction (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/assert/js/assert.js:28:13)
at SamplingField.step (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/projectile-data-lab/js/sampling/model/SamplingField.js:301:19)
at SamplingModel.step (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/projectile-data-lab/js/sampling/model/SamplingModel.js:219:30)
at http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:353:22
at PhetioAction.execute (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/tandem/js/PhetioAction.js:137:17)
at Sim.stepSimulation (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:884:31)
at Sim.stepOneFrame (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:874:12)
at Sim.runAnimationLoop (http://128.138.93.172/continuous-testing/ct-snapshots/1715368625981/chipper/dist/js/joist/js/Sim.js:852:12)
[CONSOLE] continuous-test-wrapper-error
[CONSOLE] continuous-test-wrapper-error

id: "Sparky Node Puppeteer"
Snapshot from 5/10/2024, 1:17:05 PM
```

**Comment by samreid** [comment]
This is pretty rare on CT. Showing once in the last 20x2 = 40 columns (2x multiplier is for 2x sims). @matthew-blackman and I investigated this on 3 occasions and do not have any significant leads. We do not believe the problem is likely to happen on a client machine, since nothing related to this was discovered during the QA process. We are also skeptical about a significant change to the architecture or logic, as it would risk breaking other behavior and warrant QA retesting. We last discussed that we should invite a consultant (perhaps @zepumph ?) to help us look at it with a fresh perspective.  @matthew-blackman and @zepumph how should we proceed?

**Comment by zepumph** [comment]
I don't know the issue and I don't know how to reproduce it, but I have some thoughts!


In SamplingField, you have multiple Properties that aren't stateful, one of which is sampleMeanProperty. So at a general level this assertion doesn't surprise me at all, since I don't see any guarantee that sampleMeanProperty looks or behaves appropriately after a state set. I also see that numberOfStartedSamplesProperty is also unstateful.

I can only assume that this was purposeful for some good reason, so assuming that, would you like to try out recomputing these non stateful Property values from whatever stateful data we after upon set state completion? My first pass would be this line, but perhaps it will be buggy because it also will change numberOfCompletedSamplesProperty, which is stateful so perhaps shouldn't be changed. 

```diff
Subject: [PATCH] blarg
---
Index: js/sampling/model/SamplingField.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/sampling/model/SamplingField.ts b/js/sampling/model/SamplingField.ts
--- a/js/sampling/model/SamplingField.ts	(revision 7a9b72093c43407b527510b427db662700409f45)
+++ b/js/sampling/model/SamplingField.ts	(date 1715976241831)
@@ -186,6 +186,10 @@
         this.isContinuousLaunchingProperty.value = false;
       }
     } );
+
+    Tandem.PHET_IO_ENABLED && phet.phetio.phetioEngine.phetioStateEngine.stateSetEmitter.addListener( () => {
+      this.updateComputedProperties();
+    } );
   }
 
   /**
```

Also, I'd like to note that wiring assertions into the stateSetEmitter is @pixelzoom and my favorite new state-testing tool. For example here.

https://github.com/phetsims/natural-selection/blob/fe0b0e47c8fd34d67ed1b7179e419866ae67385a/js/common/model/BunnyCollection.ts#L222-L226

Let me know if I can assist any further.

**Comment by pixelzoom** [comment]
This smells like a missing `!isSettingPhetioStateProperty.value` guard.

**Comment by pixelzoom** [comment]
I see lots of places where restoring a Property will result calling SampleField `updateComputedProperties` before `sampleMeanProperty` has been restored.

For example, in SampleField.ts:

```ts
    const phaseChanged = () => {
      this.phaseStartTimeProperty.value = this.timeProperty.value;
      this.updateComputedProperties();
    };

    this.phaseProperty.link( phaseChanged );
```

... and `updateComputedProperties` sets ` this.sampleMeanProperty.value = null` (line 205).

**Comment by matthew-blackman** [comment]
@samreid and I discussed how to track down the source of this error, and want to create a historyBuffer to see which phases are being set in which order on SamplingField when this occurs. The following patch should give us a log of the last 10 phases being set. @samreid can you review? Feel free to commit if it looks good or after updating.

<details>

```diff

Subject: [PATCH] Round values sent to CODAP, remove unused launch height code - see https://github.com/phetsims/projectile-data-lab/issues/354
---
Index: js/sampling/model/SamplingField.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/sampling/model/SamplingField.ts b/js/sampling/model/SamplingField.ts
--- a/js/sampling/model/SamplingField.ts	(revision 183b1bc6c1e3267a22761dd50fe1be3153f4a667)
+++ b/js/sampling/model/SamplingField.ts	(date 1724087547286)
@@ -75,6 +75,9 @@
 
   public readonly selectedSampleNumberProperty: NumberProperty;
 
+  // The history buffer keeps track of the last 10 phases, so we can show the user what has happened.
+  private historyBuffer: string[] = [];
+
   // When pressing the eraser button, if the isContinuousLaunchingProperty is true, the field will automatically restart
   // on the next time step. See https://github.com/phetsims/projectile-data-lab/issues/289
   private _shouldResumeAfterClear = false;
@@ -167,6 +170,10 @@
     } );
 
     const phaseChanged = () => {
+
+      // Update the history buffer with the new phase
+      this.updateHistoryBuffer( this.phaseProperty.value );
+
       this.phaseStartTimeProperty.value = this.timeProperty.value;
       this.updateComputedProperties();
     };
@@ -181,7 +188,7 @@
       this.updateComputedProperties();
 
       if ( this.phaseProperty.value === 'showingCompleteSampleWithoutMean' ) {
-        assert && assert( this.sampleMeanProperty.value !== null, 'sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase. Projectiles in selected sample: ' + this.getProjectilesInSelectedSample().length + '. Sample size: ' + this.sampleSize );
+        assert && assert( this.sampleMeanProperty.value !== null, 'sampleMeanProperty should not be null in showingCompleteSampleWithoutMean phase. Projectiles in selected sample: ' + this.getProjectilesInSelectedSample().length + '. Sample size: ' + this.sampleSize + ', Phase history buffer: ' + this.historyBuffer.join( ' -> ' ) );
       }
     } );
 
@@ -315,6 +322,16 @@
     }
   }
 
+  private updateHistoryBuffer( phase: string ): void {
+    // Add the new phase to the end of the buffer
+    this.historyBuffer.push( phase );
+
+    // If the buffer exceeds 10 elements, remove the oldest one
+    if ( this.historyBuffer.length > 10 ) {
+      this.historyBuffer.shift();
+    }
+  }
+
   public step( dt: number, timeSpeed: TimeSpeed ): void {
 
     this.stepAirborneProjectiles( dt );

**Comment by matthew-blackman** [comment]
@samreid and I figured out the source of the issue (using the history buffer as a paper trail). Here is what was happening in the state and migration wrappers to cause the error:

The error occurs when the upstream sim is in the 'showingCompleteSampleWithoutMean' phase, and the downstream sim is in the 'idle' phase and being updated by the upstream sim.

In this case the sampleMean is still null in order to prevent the histogram from updating until the mean is shown. This is the correct behavior, but ran into issues when 'step' was no longer being called to update the phase. This was causing the sampleMean non-null assertion to trigger when going to the 'showingCompleteSampleWithoutMean', when in fact it was completely okay for the sampleMean to be null in that case.

The commit above updates the sampleMean non-null assertion to check for the correct phase - 'showingCompleteSampleWithMean'. This causes the error to go away, and from our testing has no effect on the behavior of the sampling screen, histogram or meanTone.

@samreid and I thought it was best to leave the historyBuffer code in case we need to do further debugging of this or related issues. The historyBuffer only updates if assertions are enabled, and has a maximum length of 10 elements, so we feel that this is okay to keep in case future issues arise.

Let's monitor CT for a few days before closing this issue. Thanks for the help @zepumph and @pixelzoom, and nice work @samreid! This was a tough issue to track down, but became much easier to reproduce once we added the phase historyBuffer. Great stuff! Will close when ready.
